E1520 recommends an invalid combination of `@singledispatch` and `@staticmethod`

### Bug description

Hi from Ruff!

We recently received a [bug report](https://github.com/astral-sh/ruff/issues/10619) at Ruff stating that our reimplementation of E1520 was recommending an invalid combination of `@functools.singledispatch` and `@staticmethod`. For example, this code runs fine, but ruff's PLE1520 rule was telling the issue reporter that they should use `@singledispatch` here instead of `@singledispatchmethod`. If you do so, the code raise an exception at runtime, as `@staticmethod` is a descriptor, and `@singledispatch` isn't designed to be stacked on top of descriptor instances (that's what `@singledispatchmethod` is for). The Python docs also state that `@singledispatchmethod` should be used for static methods: https://docs.python.org/3/library/functools.html#functools.singledispatchmethod.

```py
from functools import singledispatch, singledispatchmethod

class Foo:
    @singledispatchmethod
    @staticmethod
    def bar(value):
        raise NotImplementedError() 

    @bar.register
    @staticmethod
    def _(value: int) -> int:
        return 42 * value

    def method(self) -> int:
        return self.bar(2)

print(Foo().method())
```

After investigating the bug, we decided to change our PLE1520 rule so that it no longer complained about `@singledispatchmethod` on top of `@staticmethod`. It looks like pylint has the same incorrect behaviour here in the original PLE1520 rule, though, so I thought I'd open an issue to give you a heads-up as well!

In case it's helpful, here's the fix that was contributed to Ruff to fix the bug: https://github.com/astral-sh/ruff/pull/10637. The fix also involved some small changes to PLE1519 (E1519 at pylint) as well as PLE1520 (E1520 at pylint).


### Command used

```shell
pylint demo.py
```


### Pylint output

```
************* Module demo
demo.py:7:35: C0303: Trailing whitespace (trailing-whitespace)
demo.py:1:0: C0114: Missing module docstring (missing-module-docstring)
demo.py:3:0: C0115: Missing class docstring (missing-class-docstring)
demo.py:6:4: C0116: Missing function or method docstring (missing-function-docstring)
demo.py:6:4: C0104: Disallowed name "bar" (disallowed-name)
demo.py:4:5: E1520: singledispatchmethod decorator should not be used with functions, use singledispatch instead. (singledispatchmethod-function)
demo.py:9:5: E1520: singledispatchmethod decorator should not be used with functions, use singledispatch instead. (singledispatchmethod-function)
demo.py:14:4: C0116: Missing function or method docstring (missing-function-docstring)
demo.py:1:0: W0611: Unused singledispatch imported from functools (unused-import)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

```
************* Module demo
demo.py:7:35: C0303: Trailing whitespace (trailing-whitespace)
demo.py:1:0: C0114: Missing module docstring (missing-module-docstring)
demo.py:3:0: C0115: Missing class docstring (missing-class-docstring)
demo.py:6:4: C0116: Missing function or method docstring (missing-function-docstring)
demo.py:6:4: C0104: Disallowed name "bar" (disallowed-name)
demo.py:14:4: C0116: Missing function or method docstring (missing-function-docstring)
demo.py:1:0: W0611: Unused singledispatch imported from functools (unused-import)

------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +0.00)
```

### Pylint version

```shell
pylint 3.1.0
astroid 3.1.0
Python 3.12.2 (main, Feb 15 2024, 19:30:27) [Clang 15.0.0 (clang-1500.1.0.2.5)]
```

### OS / Environment

MacOS; Python 3.12.2

### Additional dependencies

None

--- a/doc/whatsnew/3/3.1/index.rst
+++ b/doc/whatsnew/3/3.1/index.rst
@@ -17,6 +17,29 @@ and a smattering of bug fixes.
 
 .. towncrier release notes start
 
+What's new in Pylint 3.1.1?
+---------------------------
+Release date: 2024-05-13
+
+
+False Positives Fixed
+---------------------
+
+- Treat `attrs.define` and `attrs.frozen` as dataclass decorators in
+  `too-few-public-methods` check.
+
+  Closes #9345 (`#9345 <https://github.com/pylint-dev/pylint/issues/9345>`_)
+
+- Fix a false positive with ``singledispatchmethod-function`` when a method is decorated with both ``functools.singledispatchmethod`` and ``staticmethod``.
+
+  Closes #9531 (`#9531 <https://github.com/pylint-dev/pylint/issues/9531>`_)
+
+- Fix a false positive for ``consider-using-dict-items`` when iterating using ``keys()`` and then deleting an item using the key as a lookup.
+
+  Closes #9554 (`#9554 <https://github.com/pylint-dev/pylint/issues/9554>`_)
+
+
+
 What's new in Pylint 3.1.0?
 ---------------------------
 Release date: 2024-02-25

--- a/doc/whatsnew/fragments/9345.false_positive
+++ b/doc/whatsnew/fragments/9345.false_positive
@@ -1,4 +1,0 @@
-Treat `attrs.define` and `attrs.frozen` as dataclass decorators in
-`too-few-public-methods` check.
-
-Closes #9345

--- a/doc/whatsnew/fragments/9531.false_positive
+++ b/doc/whatsnew/fragments/9531.false_positive
@@ -1,3 +1,0 @@
-Fix a false positive with ``singledispatchmethod-function`` when a method is decorated with both ``functools.singledispatchmethod`` and ``staticmethod``.
-
-Closes #9531

--- a/doc/whatsnew/fragments/9554.false_positive
+++ b/doc/whatsnew/fragments/9554.false_positive
@@ -1,3 +1,0 @@
-Fix a false positive for ``consider-using-dict-items`` when iterating using ``keys()`` and then deleting an item using the key as a lookup.
-
-Closes #9554

--- a/pylint/__pkginfo__.py
+++ b/pylint/__pkginfo__.py
@@ -9,7 +9,7 @@ It's updated via tbump, do not modify.
 
 from __future__ import annotations
 
-__version__ = "3.1.0"
+__version__ = "3.1.1"
 
 
 def get_numversion_from_version(v: str) -> tuple[int, int, int]:

--- a/tbump.toml
+++ b/tbump.toml
@@ -1,7 +1,7 @@
 github_url = "https://github.com/pylint-dev/pylint"
 
 [version]
-current = "3.1.0"
+current = "3.1.1"
 regex = '''
 ^(?P<major>0|[1-9]\d*)
 \.

--- a/towncrier.toml
+++ b/towncrier.toml
@@ -1,5 +1,5 @@
 [tool.towncrier]
-version = "3.1.0"
+version = "3.1.1"
 directory = "doc/whatsnew/fragments"
 filename = "doc/whatsnew/3/3.1/index.rst"
 template = "doc/whatsnew/fragments/_template.rst"

